### PR TITLE
Capture build cache key in origin metadata

### PR DESCRIPTION
**Pull request by wolfs**
To use it to look up the origin work unit in scans.

### CODE REVIEW

File: platforms/core-execution/execution/src/main/java/org/gradle/internal/execution/steps/CaptureOutputsAfterExecutionStep.java [hunk 1]
```diff
@@ -80,7 +82,16 @@ private ExecutionOutputState captureOutputsAfterExecution(UnitOfWork work, C con
                 // As this is _roughly_ the amount of time that is avoided by reusing the outputs,
                 // which is currently the _only_ thing this value is used for.
                 Duration originExecutionTime = result.getDuration().plus(Duration.ofMillis(snapshotOutputDuration));
-                OriginMetadata originMetadata = new OriginMetadata(buildInvocationScopeId.asString(), originExecutionTime);
+                HashCode buildCacheKey = context.getCachingState()
+                    .fold(
+                        (CachingState.Enabled enabled) -> Optional.of(enabled.getKey()),
+                        CachingState.Disabled::getKey)
+                    .map(cacheKey -> HashCode.fromBytes(cacheKey.toByteArray()))
+                    .orElse(null);
```
Comment (lptr) [inline platforms/core-execution/execution/src/main/java/org/gradle/internal/execution/steps/CaptureOutputsAfterExecutionStep.java:90]: :x: We should never get `null` here. The problem is that the cache key does not belong to the caching state anymore after your changes. It should instead be moved to the `BeforeExecutionState`, and should be calculated there. Once we do that, it will always be available when a `BeforeExecutionState` is captured.

That would also allow `OriginMetadata.getHashCode()` to be non-null, as it should be.

File: platforms/core-execution/execution/src/main/java/org/gradle/internal/execution/steps/InputChangesContext.java [hunk 2]
```diff
@@ -21,11 +21,11 @@
 import javax.annotation.Nullable;
 import java.util.Optional;
 
-public class InputChangesContext extends ValidationFinishedContext {
+public class InputChangesContext extends CachingContext {
 
     private final InputChangesInternal inputChanges;
 
-    public InputChangesContext(ValidationFinishedContext parent, @Nullable InputChangesInternal inputChanges) {
+    public InputChangesContext(CachingContext parent, @Nullable InputChangesInternal inputChanges) {
```
Comment (lptr) [inline platforms/core-execution/execution/src/main/java/org/gradle/internal/execution/steps/InputChangesContext.java:28]: ❓ Why is this change necessary?

File: platforms/core-execution/execution/src/main/java/org/gradle/internal/execution/steps/CaptureOutputsAfterExecutionStep.java (hunk 1 above)
Comment (lptr) [reply platforms/core-execution/execution/src/main/java/org/gradle/internal/execution/steps/CaptureOutputsAfterExecutionStep.java:90]: There's one more problem with doing this, which is `SkipEmptyNonIncrementalWorkStep`. I think in the long term we should stop supporting it (maybe in 9.0). Until then we can move it after `CaptureNonIncrementalStateBeforeExecutionStep`, and thus get a build cache key captured.

File: platforms/core-execution/execution/src/test/groovy/org/gradle/internal/execution/history/impl/OriginMetadataSerializerTest.groovy [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+    def serializer = new OriginMetadataSerializer()
+
+    def "can serialize and deserialize origin metadata"() {
+        def originMetadata = new OriginMetadata("test-invocation-id", buildCacheKey, Duration.ofSeconds(123))
+
+        expect:
+        serialize(originMetadata, serializer) == originMetadata
+
+        where:
+        buildCacheKey << [HashCode.fromString("1234567890abcdef"), null]
```
Comment (lptr) [inline platforms/core-execution/execution/src/test/groovy/org/gradle/internal/execution/history/impl/OriginMetadataSerializerTest.groovy:36]: 💅 We have `TestHashCodes` for this purpose. That makes sure we actually generate the right length of `HashCode`. If you want, you could add a version with a `String` input, but I think using the `long` parameter, or even adding a method that just generates a random one, is better.

File: platforms/core-execution/execution/src/main/java/org/gradle/internal/execution/steps/CaptureOutputsAfterExecutionStep.java (hunk 1 above)
Comment (wolfs) [reply platforms/core-execution/execution/src/main/java/org/gradle/internal/execution/steps/CaptureOutputsAfterExecutionStep.java:90]: We don't always calculate the cache key when e.g. build caching is off. Though we always create the origin metadata.
You can see this from the tests as well: When build caching is off and the scan plugin isn't applied, the cache key will be null here.

File: platforms/core-execution/execution/src/main/java/org/gradle/internal/execution/steps/InputChangesContext.java (hunk 2 above)
Comment (wolfs) [reply platforms/core-execution/execution/src/main/java/org/gradle/internal/execution/steps/InputChangesContext.java:28]: If I don't do this, the chaining doesn't work any more in ExecutionServices. More precisely, then context were not a strict hierarchy, and one of the two hierarchies did not include `CachingContext`: 
image

I was actually surprised that `InputChangesContext` isn't a `CachingContext`.

**Review by sorin-florea — APPROVED** [review]

**Review by facewindu — APPROVED** [review]

File: platforms/core-execution/execution/src/main/java/org/gradle/internal/execution/steps/CaptureOutputsAfterExecutionStep.java (hunk 1 above)
Comment (wolfs) [reply platforms/core-execution/execution/src/main/java/org/gradle/internal/execution/steps/CaptureOutputsAfterExecutionStep.java:90]: Hmm... It seems that we already don't support skip when empty for non-incremental work: https://github.com/gradle/gradle/blob/c399528b58d980780c234330436d4890bd170688/platforms/software/dependency-management/src/main/java/org/gradle/api/internal/artifacts/DependencyManagementGlobalScopeServices.java#L123-L173

Am I missing something?

File: platforms/core-execution/build-cache-base/src/main/java/org/gradle/caching/internal/origin/OriginMetadataFactory.java [hunk 4]
```diff
@@ -24,6 +26,7 @@ public class OriginMetadataFactory {
     private static final String BUILD_INVOCATION_ID_KEY = "buildInvocationId";
     private static final String TYPE_KEY = "type";
     private static final String IDENTITY_KEY = "identity";
+    private static final String BUILD_CACHE_KEY_KEY = "buildCacheKey";
```
[the remark this answers is not in the document]
Comment (wolfs) [reply platforms/core-execution/build-cache-base/src/main/java/org/gradle/caching/internal/origin/OriginMetadataFactory.java]: Done.

File: platforms/core-execution/build-cache-base/src/main/java/org/gradle/caching/internal/origin/OriginMetadataFactory.java [hunk 5]
```diff
@@ -50,12 +53,13 @@ public OriginMetadataFactory(
         this.hostnameLookup = hostnameLookup;
     }
 
-    public OriginWriter createWriter(String identity, Class<?> workType, Duration elapsedTime) {
+    public OriginWriter createWriter(String identity, Class<?> workType, String buildCacheKey, Duration elapsedTime) {
```
[the remark this answers is not in the document]
Comment (wolfs) [reply platforms/core-execution/build-cache-base/src/main/java/org/gradle/caching/internal/origin/OriginMetadataFactory.java]: Done.

File: platforms/core-execution/execution/src/main/java/org/gradle/internal/execution/steps/CaptureOutputsAfterExecutionStep.java (hunk 1 above)
Comment (wolfs) [reply platforms/core-execution/execution/src/main/java/org/gradle/internal/execution/steps/CaptureOutputsAfterExecutionStep.java]: Alright, now the `cacheKey` is non-null in `OriginMetadata` 🎉

File: platforms/core-execution/execution/src/test/groovy/org/gradle/internal/execution/history/impl/OriginMetadataSerializerTest.groovy (hunk 3 above)
Comment (wolfs) [reply platforms/core-execution/execution/src/test/groovy/org/gradle/internal/execution/history/impl/OriginMetadataSerializerTest.groovy]: Using `TestHashCodes` now.

File: platforms/core-execution/build-cache-base/src/main/java/org/gradle/caching/internal/origin/OriginMetadata.java [hunk 6]
```diff
@@ -64,6 +79,7 @@ public int hashCode() {
     public String toString() {
         return "OriginMetadata{"
             + "buildInvocationId=" + buildInvocationId
+            + "buildCacheKey=" + buildCacheKey
```
[the remark this answers is not in the document]
Comment (wolfs) [reply platforms/core-execution/build-cache-base/src/main/java/org/gradle/caching/internal/origin/OriginMetadata.java]: Fixed.

File: platforms/core-execution/build-cache-base/src/main/java/org/gradle/caching/internal/origin/OriginMetadata.java [hunk 7]
```diff
@@ -16,22 +16,30 @@
...
         this.executionTime = executionTime;
     }
 
     public String getBuildInvocationId() {
         return buildInvocationId;
     }
 
+    public HashCode getBuildCacheKey() {
+        return buildCacheKey;
+    }
```
[the remark this answers is not in the document]
Comment (wolfs) [reply platforms/core-execution/build-cache-base/src/main/java/org/gradle/caching/internal/origin/OriginMetadata.java:41]: Yes, that is true. If that is the reason we don't use `BuildInvocationScopeId` here, then I suppose we may want to use `String` here as well.

File: platforms/core-execution/execution/src/main/java/org/gradle/internal/execution/steps/CaptureOutputsAfterExecutionStep.java (hunk 1 above)
Comment (lptr) [reply platforms/core-execution/execution/src/main/java/org/gradle/internal/execution/steps/CaptureOutputsAfterExecutionStep.java]: Yay!

File: platforms/core-execution/execution/src/main/java/org/gradle/internal/execution/steps/InputChangesContext.java (hunk 2 above)
Comment (lptr) [reply platforms/core-execution/execution/src/main/java/org/gradle/internal/execution/steps/InputChangesContext.java]: I remember that being intentional. I wanted to go towards a more independent, less linear-looking context hierarchy. But it never worked. I think it's fine to change it the way you did. We should come back to this with generated context implementations sometime later.

File: platforms/core-execution/execution/src/main/java/org/gradle/internal/execution/steps/AbstractResolveCachingStateStep.java [hunk 8]
```diff
@@ -67,12 +67,17 @@ public CachingResult execute(UnitOfWork work, C context) {
         return new CachingResult(result, cachingState);
     }
```
Comment (lptr) [inline platforms/core-execution/execution/src/main/java/org/gradle/internal/execution/steps/AbstractResolveCachingStateStep.java:68]: :x: Let's move the logging to where it belongs.

File: platforms/core-execution/execution/src/main/java/org/gradle/internal/execution/steps/CaptureOutputsAfterExecutionStep.java [hunk 9]
```diff
@@ -80,7 +82,16 @@ private ExecutionOutputState captureOutputsAfterExecution(UnitOfWork work, C con
                 // As this is _roughly_ the amount of time that is avoided by reusing the outputs,
                 // which is currently the _only_ thing this value is used for.
                 Duration originExecutionTime = result.getDuration().plus(Duration.ofMillis(snapshotOutputDuration));
-                OriginMetadata originMetadata = new OriginMetadata(buildInvocationScopeId.asString(), originExecutionTime);
+                HashCode buildCacheKey = context.getCachingState()
+                    .fold(
+                        (CachingState.Enabled enabled) -> Optional.of(enabled.getKey()),
+                        CachingState.Disabled::getKey)
+                    .map(cacheKey -> HashCode.fromBytes(cacheKey.toByteArray()))
```
[the remark this answers is not in the document]
Comment (wolfs) [reply platforms/core-execution/execution/src/main/java/org/gradle/internal/execution/steps/CaptureOutputsAfterExecutionStep.java]: This code is gone.

File: platforms/core-execution/execution/src/main/java/org/gradle/internal/execution/steps/AbstractResolveCachingStateStep.java (hunk 8 above)
Comment (wolfs) [reply platforms/core-execution/execution/src/main/java/org/gradle/internal/execution/steps/AbstractResolveCachingStateStep.java:68]: As discussed, the logging seems to be fine as is.

File: platforms/core-execution/execution/src/main/java/org/gradle/internal/execution/caching/CachingState.java [hunk 10]
```diff
@@ -90,18 +94,39 @@ public BeforeExecutionState getBeforeExecutionState() {
...
+            return cacheKeyCalculatedState;
+        }
+
+        public BuildCacheKey getKey() {
+            return cacheKeyCalculatedState.getKey();
+        }
+
+        public BeforeExecutionState getBeforeExecutionState() {
+            return cacheKeyCalculatedState.getBeforeExecutionState();
+        }
```
[the remark this answers is not in the document]
Comment (wolfs) [reply platforms/core-execution/execution/src/main/java/org/gradle/internal/execution/caching/CachingState.java]: Removed.

File: platforms/core-execution/execution/src/main/java/org/gradle/internal/execution/caching/CachingState.java [hunk 11]
```diff
@@ -111,20 +136,8 @@ public ImmutableList<CachingDisabledReason> getDisabledReasons() {
...
-        public Optional<BuildCacheKey> getKey() {
-            return Optional.ofNullable(key);
-        }
-
-        /**
-         * Individual fingerprints for each of the work's inputs, if they were tracked.
-         */
-        public Optional<BeforeExecutionState> getBeforeExecutionState() {
-            return Optional.ofNullable(beforeExecutionState);
+        public Optional<CacheKeyCalculatedState> getCacheKeyCalculatedState() {
```
[the remark this answers is not in the document]
Comment (wolfs) [reply platforms/core-execution/execution/src/main/java/org/gradle/internal/execution/caching/CachingState.java:142]: I added some Javadoc.

File: platforms/core-execution/build-cache-base/src/main/java/org/gradle/caching/internal/origin/OriginMetadata.java (hunk 7 above)
[the remark this answers is not in the document]
Comment (wolfs) [reply platforms/core-execution/build-cache-base/src/main/java/org/gradle/caching/internal/origin/OriginMetadata.java:41]: We discussed that `HashCode` is a core type that is fine to expose in these code areas.

**Review by lptr — APPROVED** [review]
LGTM!

**Comment by wolfs** [comment]
@bot-gradle merge
